Add a feature to parse SubjectPublicKeyInfo into (signature::Algorithm, SubjectPublicKey)

marshallpierce opened the pull request:
Not yet ready for merging. I wanted to start adding notes and questions, though.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -472,154 +449,139 @@ mod tests {
-        Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithm));
-    test_verify_signed_data_signature_outer!(
-        test_ecdsa_prime256v1_sha512_unused_bits_signature,
-        "ecdsa-prime256v1-sha512-unused-bits-signature.pem",
```
Test deleted: the signature DER is invalid, so not relevant for signature validation since we can't decode the signature itself.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -472,154 +449,139 @@ mod tests {
-        Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithm));
-    test_verify_signed_data!(
-        test_ecdsa_prime256v1_sha512_using_rsa_algorithm,
-        "ecdsa-prime256v1-sha512-using-rsa-algorithm.pem",
```
Test deleted: error is in Algorithm field, but that's not used for plain signature validation (only SignedData).

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -472,154 +449,139 @@ mod tests {
+        &spki::RSA_PKCS1_2048_8192_SHA1,
         Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithmForPublicKey));
-    test_verify_signed_data!(test_rsa_pkcs1_sha1_wrong_algorithm,
-                             "rsa-pkcs1-sha1-wrong-algorithm.pem",
```
Test deleted: error is in Algorithm, which we don't use

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -472,154 +449,139 @@ mod tests {
+                           &spki::RSA_PKCS1_2048_8192_SHA256,
+                           Err(VerifyWithSPKIError::InvalidSignatureForPublicKey));
+    // TODO this only tests DER decoding, not signature logic
     test_parse_spki_bad_outer!(test_rsa_pkcs1_sha256_key_encoded_ber,
```
Should these parse_spki_bad_outer tests be moved to the DER module? It doesn't really have anything to do with signatures.

briansmith replied:
The existence of those tests in webpki is just an artifact of history, really. I think we should just leave them all out of the *ring* test suite.

marshallpierce replied:
Works for me. I'll check to make sure that functionality has some tests but if so I'll delete these.

marshallpierce replied:
There weren't tests that I saaw for that `der` method so I moved those tests over. This involved some mildly unpleasant compromises: I moved `read_pem_section` over to `der::tests` and used it from there for `spki`'s testing needs. I felt it was worth it to not lose out on some real-world-ish test coverage for the various flavors of incorrect DER expressed by those tests, but I'll ditch 'em if you'd rather.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -472,154 +449,139 @@ mod tests {
-    test_verify_signed_data!(
-        test_rsa_pkcs1_sha256_using_ecdsa_algorithm,
-        "rsa-pkcs1-sha256-using-ecdsa-algorithm.pem",
-        Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithmForPublicKey));
```
Test deleted: error in Algorithm

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -472,154 +449,139 @@ mod tests {
         Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithmForPublicKey));
 
-    // Chromium's PSS test are for parameter combinations we don't support.
-    test_verify_signed_data!(test_rsa_pss_sha1_salt20_using_pss_key_no_params,
```
PSS tests deleted: no point in a bunch more tests that only check that 1024 bit keys aren't supported...

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,488 @@
+/// A signature algorithm.
+pub struct SignatureAlgorithm {
+    public_key_alg_id: AlgorithmIdentifier,
+    signature_alg_id: AlgorithmIdentifier,
```
This field is now unused since we don't need to select algorithms by digest. What do you think is the right abstraction? Should we remove this field (and all the associated constants)? Leave it, but document its use in SignedData? 
It seems good to have *some* way to represent signature info, but I'm not sure how widely applicable those signature alg OIDs are.

briansmith replied:
It seems like the right thing to remove `signature_alg_id`. That means that webpki would have to continue to define all of these things, as a wrapper around `ring::signature::spki::SignatureAlgorithm` and `signature_alg_id`.

briansmith replied:
It might be a good idea to rename `SignatureAlgorithm` to `AlgorithmIdentifier`, leaving the `Signature` part implied by the fact that it is nested in `ring::signature`.

marshallpierce replied:
OK, I'll work on that. A shame that "Signature" is overloaded both to be the general concept, the noun, and also the creation (as opposed to verification) of the thing...

Also, ICYMI, I made a few comments on an individual commit as I was massaging the tests that show as outdated.

marshallpierce replied:
`AlgorithmIdentifier` was already claimed in `spki` so I went with `Algorithm`.

briansmith commented on src/der.rs:
```diff
@@ -168,6 +168,19 @@ pub fn small_nonnegative_integer(input: &mut untrusted::Reader)
 
+pub fn bit_string_with_no_unused_bits<'a>(input: &mut untrusted::Reader<'a>)
+                                          -> Result<untrusted::Input<'a>,
+                                              error::Unspecified> {
```
I filed #379 for this.

briansmith commented:
Note that in 0.7.x I'll be moving lots (most? almost all) of *ring* tests to tests/ because they're really integration tests. In particular, all tests that only use the public API will move to tests/. It seems like it would be great to do this from the beginning for this new feature.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -472,154 +449,139 @@ mod tests {
 
     // XXX: Some of the BadDER tests should have better error codes, maybe?
 
+    // TODO need ECDSA_P256_SHA512 for these tests
```
I was planning on scripting a few `openssl` invocations to generate keys of all the types in `spki` and all corresponding digests, generating sample signatures, and then writing tests to verify that signatures match/don't match as appropriate so that every algorithm/curve/digest flavor has at least some coverage. Open to suggestions, though!

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,197 @@
+// with SHA-1) say that parameters must be omitted. RFC 4055 Section 5 and RFC
+// 3279 Section 2.2.1 both say that parameters for RSA must be encoded as NULL;
+// we relax that requirement by allowing the NULL to be omitted, to match all
+// the other signature algorithms we support and for compatibility.
```
Is this comment still relevant? The RSA_ENCRYPTION alg id has this structure:

```
    0:d=0  hl=2 l=   9 prim: OBJECT            :rsaEncryption
   11:d=0  hl=2 l=   0 prim: NULL
```

It looks like we do require the NULL...

briansmith replied:
I believe that comment is referring to *signature* algorithm identifiers, not *key* algorithm identifiers, so it won't apply in any case since we remove `signature_alg_id` in a later commit. And yes, it appears to be outdated. I just removed it in webpki: https://github.com/briansmith/webpki/commit/2419bb730d0c8d2f315a4a38eea44d601428272d and it should be removed here as well.

briansmith replied:
It was out of date. I removed the comment in https://github.com/briansmith/webpki/commit/2419bb730d0c8d2f315a4a38eea44d601428272d.

marshallpierce replied:
OK, removed.

marshallpierce commented on tests/test-data/gen-test-signatures.sh:
```diff
@@ -0,0 +1,71 @@
+#!/bin/sh
```
Let me know what you think of this approach. I wanted to have some simple sanity checks that basic signature validation for all the supported algorithm combinations, and this seemed like a simple way to get that (assuming openssl's output is correct, of course).

marshallpierce commented on tests/test-data/gen-test-signatures.sh:
```diff
@@ -0,0 +1,71 @@
+
+messages=( msg1 msg2 )
+rsa_key_lengths=( 2048 4096 8192 )
+rsa_padding_algs=( pkcs )
```
openssl's CLI doesn't support PSS signatures AFAICT but there's some C in http://openssl.6102.n7.nabble.com/RSASSA-PSS-Signature-Question-td22572.html that should be a good start to making a little single-purpose tool we can use for generating signatures.

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,615 @@
+#[cfg(feature = "rsa_signing")]
+use ring::rand;
+
+// TODO: The expected results need to be modified for SHA-1 deprecation.
```
This was in the original tests; I'm not sure what it means at this point

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,615 @@
+        #[test]
+        fn $fn_name() {
+            // check existing signature file with spki pub key
+            rsa_verify_signature_file_with_spki($algorithm,
```
This is using the various inputs, signatures, etc generated by the `gen-test-signatures.sh` script to test success and failure modes for a given mode.

briansmith replied:
Your comment in the PR about this using the gen-test-signature.sh script output as the inputs for the tests should be a comment in this file somewhere, perhaps at the top of the file.

marshallpierce replied:
Done

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,615 @@
+                 signature));
+}
+
+macro_rules! test_rsa_sign_verify_spki {
```
No ECDSA signing support yet, so we don't have an equivalent ECDSA "sign then verify the signature" scenario.

briansmith replied:
Are the `rsa_sign_and_verify_spki` tests essential for covering the code? Normally we don't add this kind of test because, for example, we want all the verification tests to run even when RSA signing isn't enabled. Or, in other words, if this test is testing something important about signature verification then we need to find a way to do it even when RSA signing is disabled and also for ECDSA and EdDSA, but if it testing something important about signature verification then it seems like we should remove it and/or defer it. WDYT?

marshallpierce replied:
I don't think the tests are *essential*. The other tests verify signatures generated with openssl's CLI, which does work and is certainly valuable from an interop perspective. However, it felt comforting to verify that using SPKI pub keys also works with signatures that *ring* generates, given that RSA signature generation is probabilistic so *ring* will not generate the same signatures as the openssl ones. I'd also be happy to add support for the other algorithms, but until ECDSA signing is supported, RSA is as far as I can go. For EdDSA, `spki.rs` doesn't currently have any algorithms for that suite; is that planned/desired?

Anyway, if the somewhat vague value proposition is not compelling, it wouldn't break my heart to remove the rsa sign-and-verify tests.

briansmith replied:
> Anyway, if the somewhat vague value proposition is not compelling, it wouldn't break my heart to remove the rsa sign-and-verify tests.

I suggest we pull them out for now and maybe put them in another PR so we can measure the code coverage without them.

marshallpierce replied:
OK, the tests are in `spki-rsa-sign-and-verify-tests` for posterity.

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,615 @@
+                               "sha1");
+
+// 8192 bit rsa in 2048_8192 modes
+// TODO loading 8192 key pairs isn't working
```
`signature::RSAKeyPair::from_der` fails when loading this 8192-bit key.

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,615 @@
+
+type FileLines<'a> = std::io::Lines<std::io::BufReader<&'a std::fs::File>>;
+
+fn read_pem_section(lines: & mut FileLines, section_name: &str)
```
copy-and-paste of equivalent code in `der` tests

briansmith replied:
I think I understand that we need to copy/paste because these integration tests need to be run against the production version of *ring*, and we don't want these common functions to be part of the public *ring* API. Let's add a comment explaining this to this file.

Another way to do this, which I prefer, would be to put the common logic in a separate source file, e.g. tests/der_tests_common.rs, and then `include!` that file in `ring::der::tests` and in this file. WDYT?

Otherwise, we at least need to add some comments "Keep this in sync with the copy in <other file's relative path>" to both src/der.rs and this file, for each item that is duplicated.

marshallpierce replied:
I like the `include!` approach. While I'm in there I might as well solve https://github.com/briansmith/ring/issues/452 because it will be trivial.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,197 @@
+/// with `-pubout` and friends, you will get a DER of an SPKI wrapper containing
+/// a public key. You could extract the bitstring of the key itself and use
+/// `signature:;verify`, but that is often inconvient.
+pub fn verify(signature_alg: &Algorithm,
```
It looks like in webpki it was convenient to carry around just the inner two components of SPKI, but the file that people would tend to load from an openssl exported key would start with the top level sequence. Both seem like pretty valid things to loosely call "SPKI". Thoughts on how to handle this? Two functions? Helper function to strip the outer layer of DER?

briansmith replied:
If I understand you correctly, the current webpki strips the outermost SEQUENCE in a higher-level function and saves the insides, and only the inside parts are passed to the verification function. However, you're also saying *ring*'s new verification function that takes SPKIs needs to strip the SEQUENCE itself. It would be nice if webpki could pass the entire SEQUENCE to the *ring* verification function so that two different functions aren't needed. Am I understanding things correctly?

If so, the webpki code can be changed to something like this:
```rust
...
let spki_begin reader.mark();
try!(der::nested(... der::SEQUENCE, |spki| Ok(())));
let spki_end = reader.mark();
let spki = try!(reader.get_input_between_marks(spki_begin, spki_end));
...
```

Then `spki` would contain the outer later of the sequence. In fact, this logic should probably be encapsulated in `ring::der` as a function called something like `expect_tag_and_get_tlv()`.

Then webpki could just use the public *ring* function, IIUC. WDYT?

marshallpierce replied:
I think we're on the same page, but just to be sure I'll elaborate.

Tracing the data flow up from `verify_signed_data`:
```
EndEntityCert::verify_is_valid_tls_server_cert // top of call stack, takes a &[TrustAnchor] 
verify_cert::build_chain // takes a &[TrustAnchor] 
verify_cert::check_signatures // called on TrustAnchor.spki
signed_data::verify_signed_data 
signed_data::verify_signature // function now in ring
```

And `TrustAnchor` is created in `trust_anchor_util::parse_cert_v1` with an `spki` of:

```
let spki =
                    try!(der::expect_tag_and_get_value(tbs,
                                                       der::Tag::Sequence));
```

or via `trust_anchor_from_cert` in `parse_cert_internal` with identical code.

webpki's representation of spki as just the contents of the sequence is sensible because it saves a few mostly-meaningless bytes, especially when all the code that uses it is internal. 

I think the choices are as follows:

1. Leave webpki's representation as it is, and have *ring*'s function continue to expect that as input. For users that haven't already stripped off the outer sequence (as when they're loading a key from a file), document the verification function to explain that `expect_tag_and_get_value` should be called first on the DER.
1. Switch webpki to keep the complete SPKI DER internally, and switch *ring*'s function accordingly. Now both webpki and average joes working with a public key have no intervening steps to process the key data from the point of loading the bytes. This is nice cleanliness-wise, but costs a few bytes of memory to hang onto the key, and a few cycles for every verification to strip off the sequence. 

Ironically given how I started this feature request, I'm slightly inclined towards the former choice given the low-level nature and performance sensitivity of this library, but the latter is also pleasingly tidy. Either seems quite defensible, though. I *think* you're suggesting going with the second option; is that right?

briansmith commented:
> I think you're suggesting going with the second option; is that right?

Yes. If people are building other DER-parsing things, they might have to do the awkward thing I propose for webpki, but we could just make that less awkward in the future, if so.

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,700 @@
+    return vec;
+}
+
+// TODO sha1 support?
```
Is it planned to support RSA SHA1 signatures? If not I'll remove the todo.

briansmith replied:
We do support verifying RSA PKCS#1 SHA-1 signatures but not RSA PSS SHA-1 signatures. We don't support *signing* with SHA-1 at all.

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,700 @@
+                               "sha384",
+                               "rsa_2048",
+                               "sha1");
+// TODO PSS for 3072-8192?
```
Is it planned to expose PSS modes for the 3072-8192bit key RSA algorithm? If not I'll remove this todo.

briansmith replied:
Go ahead and remove that comment. I am not sure what we're doing with the `3072_8192` stuff yet.

marshallpierce commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,700 @@
+                              "ecdsa_prime256v1",
+                              "sha256");
+
+// XXX: Some of the BadDER tests should have better error codes, maybe?
```
From here on are tests that existed already in webpki's version of this code

briansmith commented on src/lib.rs:
```diff
@@ -48,7 +48,6 @@
 )]
 #![deny(
     const_err,
-    dead_code,
```
Note to self: make sure this was reverted.

briansmith commented on Cargo.toml:
```diff
@@ -262,6 +262,9 @@ name = "ring"
 untrusted = "0.3.2"
 
+[dev-dependencies]
+rustc-serialize = "0.3.15"
```
I filed #452 about replacing this dependency with a better one later.

briansmith commented on src/der.rs:
```diff
@@ -168,6 +168,19 @@ pub fn small_nonnegative_integer(input: &mut untrusted::Reader)
+                                              error::Unspecified> {
+    nested(input, Tag::BitString, error::Unspecified, |value| {
+        let unused_bits_at_end =
+        try!(value.read_byte().map_err(|_| error::Unspecified));
```
Indent the `try!` one more level.

briansmith replied:
Fix the indention by indenting the `try!` one more level.

marshallpierce replied:
Done

briansmith commented on src/der.rs:
```diff
@@ -168,6 +168,19 @@ pub fn small_nonnegative_integer(input: &mut untrusted::Reader)
+pub fn bit_string_with_no_unused_bits<'a>(input: &mut untrusted::Reader<'a>)
+                                          -> Result<untrusted::Input<'a>,
+                                              error::Unspecified> {
+    nested(input, Tag::BitString, error::Unspecified, |value| {
```
Note to self: reformat this, both the return value and the `try!` indention below.

briansmith commented on src/der.rs:
```diff
@@ -178,10 +191,13 @@ pub fn positive_integer<'a>(input: &mut untrusted::Reader<'a>)
     use error;
     use super::*;
     use untrusted;
+    use rustc_serialize::base64::FromBase64;
```
Let's sort these alphabetically.

marshallpierce replied:
Done

marshallpierce replied:
By the way, what's your preference on sorting things like `super` or `std` separately from stuff in this crate or other crates like `untrusted` or `signature?

briansmith replied:
I don't care. It looks like it's not done consistently in the existing code, which is unfortunate, but probably inevitable unless/until we have a tool enforce it.

briansmith commented on src/der.rs:
```diff
@@ -197,6 +213,57 @@ mod tests {
+
+        base64.from_base64().unwrap()
+    }
+
```
NIT: Please put take the added code that is currently above this comment, and move it to be below test_parse_bad_spki_der. This way, the code is sorted top-down.

marshallpierce replied:
That code's gone; problem solved!

briansmith commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+///
+/// A common situation where this encoding is encountered is when using public keys
+/// exported by OpenSSL. If you export an RSA or ECDSA public key from a keypair
+/// with `-pubout` and friends, you will get DER of an SPKI wrapper containing
```
s/DER of an SPKI wrapper containing a public key/DER-encoded `SubjectPublicKeyInfo`/.

marshallpierce replied:
Done

briansmith commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+/// exported by OpenSSL. If you export an RSA or ECDSA public key from a keypair
+/// with `-pubout` and friends, you will get DER of an SPKI wrapper containing
+/// a public key. You could extract the bitstring of the key itself and use
+/// `signature::verify`, but that is often inconvient.
```
Let's remove the “You could...” sentence. (Also, “inconvenient” was misspelled.)

marshallpierce replied:
Done

briansmith commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+/// with `-pubout` and friends, you will get DER of an SPKI wrapper containing
+/// a public key. You could extract the bitstring of the key itself and use
+/// `signature::verify`, but that is often inconvient.
+pub fn verify(signature_alg: &Algorithm,
```
The purpose of the `AlgorithmIdentifier` field in the SPKI is to allow the implementation to dynamically dispatch on the identifier. But this API doesn't support that.

Instead, I think we need to change the `signature_alg` parameter to `supported_algorithms: &[&'static Algorithm]`.  Then the function should loop through the algorithms in order until it finds a match on the `AlgorithmIdentifier`, and then verify the signature using that algorithm. Then we could also change the return type to be `Result<&'static Algorithm, VerifyWithSPKIError>`. This way, the caller can control exactly what algorithms are supported and they don't have to implement any dynamic dispatch logic themselves.

marshallpierce replied:
Is it common that users would have some public key ASN.1 but not know what kind of key it was? That's not my use case but I don't claim to be representative.

It also complicates error handling somewhat. We would need an implicit priority of errors: presumably a `UnsupportedSignatureAlgorithmForPublicKey` would be suppressed, *unless* the last supplied algorithm produced that error, *or* an algorithm produced `InvalidSignatureForPublicKey`.

Anyway, for users that do have a bag o' algorithms that they wish to support, suppose we have this prelude:

```
let algorithms = &[ECDSA_P384_SHA384, ...];
let pub_key = ...;
let msg = ...;
let signature = ...;
```

Is the API change providing anything more than avoiding the `iter()` and `any()` here?

```
let got_a_live_one = 
    algorithms.iter().any(|a| spki::verify(a, pub_key, msg, signature).is_ok());
```

Or, if you wanted the algorithm, `filter()` and `nth(0)` instead of `any()`. To me, this seems like not an onerous burden for the consumer, and the intent is pretty evident I think. While it would be a little bit less code with the proposed changes, it would also require the user to look up API docs to determine how the `supported_algorithms` parameter is used (first match wins?) and how error handling works, whereas when the iteration is visible the semantics are pretty clear.

Anyway, I'll do it if you want, but IMO the syntactic win is not worth the semantic complexity.

briansmith replied:
> Is it common that users would have some public key ASN.1 but not know what kind of key it was? That's not my use case but I don't claim to be representative.

I was thinking it would be common but now I recall that it may not be. For example, webpki doesn't need such a facility because it maps every signature `AlgorithmIdentifier` to a public key `AlgorithmIdentifier`. In general, every signature format should identify the signature algorithm as part of the signature, so such a mapping should always be possible for well-specified protocols. So please just ignore this change request.

briansmith commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,701 @@
+extern crate ring;
```
Eventually we'll put lots of tests in tests/ and it would be nice to have some structure to the naming. Let's rename the file to `signature_spki_tests.rs`.

Also, please copy the license block to this file.

marshallpierce replied:
Done

briansmith commented on tests/spki_tests.rs:
```diff
@@ -0,0 +1,701 @@
+    }
+}
+
+macro_rules! test_rsa_verify_sig_file_spki {
```
For the `sig_file` tests, instead of having macros and functions specific to RSA and ECDSA, let's refactor things so that we have ones that are generic to both. In particular, I think you can just rename the input files a bit so that they have “RSA_PKCS1_2048_8192_SHA256” and “ECDSA_P256_SHA256” in the name. Then you should be able to eliminate this code duplication because callers could pass `"RSA_PKCS1_2048_8192_SHA256"` instead of separate `"rsa_2048", "pkcs1", "sha256"` parameters, and they could pass `"ECDSA_P256_SHA256"` instead of separate `"ecdsa_prime256v1", "sha256"` parameters. In general, the *ring* API is designed to make it possible to write such generic code. If it's problematic to do this, let me know and I'll work through it in detail with you.

marshallpierce replied:
That's straightforward for the case where the signature should succeed, but less so for the other failure cases covered by the macro (using the other input file, a different digest algorithm, and the other key) since there isn't an easy way to find an equivalent `Algorithm` with, say, a different digest. How would you suggest handling those tests?

briansmith commented on tests/test-data/gen-test-signatures.sh:
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
```
I skipped reviewing this file for now. I guess I'll probably review it mostly indirectly by looking at what it outputs and how the tests use the outputs.

briansmith commented on tests/test-data/keys/ecdsa_prime256v1.pem:
```diff
@@ -0,0 +1,5 @@
+MHcCAQEEIKE0vihID+ffiAB2O8DW7d9kTRcGHA7a7bsr0lu2NPaKoAoGCCqGSM49
+AwEHoUQDQgAEc563pICgt7T27Hu1nnJ1Qis6jwIEx0ixBihKkeJRurqz/2EZtpmi
+o263hkYfxpgr2T0dvqu/qfSS5Jbc14Iw3A==
+-----END EC PRIVATE KEY-----
```
There's no reason for the key to be in PEM format, is there? We tolerate the Chromium tests being in PEM format because we want to keep them in sync with Chromium's tree, but for our own test vectors I'd prefer to avoid unnecessary base64 encoding. OpenSSL tools always (AFAICT) provide a way to output binary files instead of PEM, usually with `-outform der` or similar.

marshallpierce replied:
The issue was with `openssl dgst`, not the key generation, but this prompted me to go back and look and sure enough it takes a `keyform` that I missed before. `openssl genrsa` doesn't seem to allow selecting the output format, but that's what `rm` is for...

briansmith commented on third-party/chromium/data/verify_signed_data/annotate_test_data.py:
```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/python
```
I originally kept this script in the webpki repo (inherited from Chromum's tree) because I thought we might make our own tests using the same format. But instead of using that format, you made your own format to extend the tests. That is OK, but if we're not going to add our own tests using the Chromium format then we should just avoid adding annotate_test_data.py here.

Also, if we're going to use a different format for new tests, we should probably follow up this PR with another PR that converts all the Chromium test vectors into our new preferred format, so that we don't have two formats. (I don't have strong feelings about using the Chromium format or not. I originally used the Chromium tests because I was in a hurry to get tests of any kind written, and because I thought I might some day reuse their entire test suite in webpki. But that's kind of outdated thinking at this point.)

marshallpierce replied:
I don't think that the stuff generated by `gen-test-signatures.sh` is really suitable to take over the job of the Chromium test files. It was the most expedient way I could think of to provide a cross-product of signature algorithm, key size, and digest algorithm for a fixed set of inputs. The Chromium test files, on the other hand, are much more flexible, and they allow for comments easily, they are clumsy to assemble en masse (as would be needed for the above cross-product). The `tests/test-data` scheme doesn't have a good way to express things like this:

> This uses the same test data as ecdsa-prime256v1-sha512.pem, HOWEVER the
algorithm OID for the public key has been changed from id-ecPublicKey
(1.2.840.10045.2.1) to id-ecMQV (1.3.132.1.13). This test should fail because the public key's algorithm does not match that of
the signature algorithm.

but `ecdsa-prime256v1-sha512-using-ecmqv-key.pem` does that pretty well.

I'm not vigorously in favor of the Chromium format per se, but it also seemed like overkill for what I wanted to do with SPKI. I don't really know what best practices are for maintaining test files of delicately set up binary inputs and outputs. If the Chromium team has already been around this particular block and decided that their PEM format is The Right Choice, then I can re-work the spki tests to use that format. Or, we can come up with yet another format that supersedes both.

(Also, could just reimplement the Python in Rust and have it live in `examples` as a dev tool.)

briansmith commented:
Also, please add the contributor's statement to the end of each commit: https://github.com/briansmith/ring#contributing. Also, are you the copyright holder for your changes, or is somebody else (e.g., an employer)? If not you, who?

Thanks!

marshallpierce commented:
I am the copyright holder for all of my changes.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+/// `signature::verify` is suitable when the public key bytes you have are the
+/// DER ASN.1 of the key. This method is suitable when the public key is in
+/// SubjectPublicKeyInfo form (https://tools.ietf.org/html/rfc5280#section-4.1),
+/// which is a wrapper containing both the key and an identifier.
```
Done

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+    UnsupportedSignatureAlgorithmForPublicKey,
+}
+
+/// Verify the signature `signature` of message `msg` with the public key
```
Done

marshallpierce commented:
The GitHub UI gets pretty unwieldy with PRs with a lot of comments like this, so I went through the hidden "outdated" comments and came up with this list of things still pending:

- sha 1 deprecation comment? https://github.com/briansmith/ring/pull/378#discussion_r93111120
- test organization / macro usage https://github.com/briansmith/ring/pull/378#discussion_r99931474
- test file format https://github.com/briansmith/ring/pull/378#discussion_r99938597

Oh how I wish for BitBucket's "tasks".

briansmith commented:
tl;dr: I think when we made the plan for this a while back, we decided too early to move the internal webpki verification API to *ring*, and I think we can do something better and simpler.

@marshallpierce Thanks for your very interesting response to my suggestion. to replace the `alg: &[Algorithm]` parameter in the new `verify()` with an `supported_algs: &[&'static Algorithm]` parameter. It really got me thinking about what should we ask the caller to do, and what should be do for the caller automatically. I think your reasoning can be extended nicely addresses not only the API orthogonality issues in that subthread, but also even bigger issues.

First, consider this refactoring of the function that we're adding:
```
pub fn verify(alg: &Algorithm,
              public_key_spki: untrusted::Input,
              msg: untrusted::Input,
              signature: untrusted::Input) -> Result<(), VerifyWithSPKIError> {
    let public_key = try!(parse_spki(alg, public_key_spki, signature));
    super::verify(alg, public_key, msg, signature)
}

```rust
fn parse_spki(alg: &Algorithm,
              public_key_spki: untrusted::Input,
              msg: untrusted::Input,
              signature: untrusted::Input) -> Result<(), VerifyWithSPKIError> {
    let unwrapped_spki_der = public_key_spki.read_all(VerifyWithSPKIError::BadDER, |input| {
        der::expect_tag_and_get_value(input, der::Tag::Sequence)
            .map_err(|_| VerifyWithSPKIError::BadDER)
    }).unwrap();

    let spki = try!(parse_spki_value(unwrapped_spki_der));
    if !signature_alg.public_key_alg_id
        .matches_algorithm_id_value(spki.algorithm_id_value) {
        return Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithmForPublicKey);
    }
}
```

Note that this is just splitting out the parsing and AlgorithmIdentifier OID matching into a separate function.

Now, let's return to your reasoning above, where you say that it is better to accept a single `Algorithm` as input instead of looping through a sequence of them, for the good reasons you listed.

For similar reasons, wouldn't it be better for us to make the above `parse_spki()` function public, remove `ring::signature::spki::verify()`, and then have the caller do the `ring::signature::parse_spki(); ring::signature::verify()` steps themselves?

We have a `verify()` wrapper around `ring::signature::verify()` in webpki because webpki *always* verifies signatures where the key is encoded in an SPKI; in other words, that wrapper is just factoring out the  *common* `parse_spki(); verify()` logic in webpki. But that sequence isn't common in every code base using *ring*, which is why we have been planning to keep `ring::signature::verify()` and add `ring::signature::spki::verify()` alongside it. But that itself is a problem because now we have to explain to users of *ring* how to choose between two different `verify()` functions.

Further, if we did replace the new `verify()` function with a `parse_spki()` function then the issues we've encountered with the testing—e.g. how to deal with algorithms we don't support, how we deal with the duplication between the Chromium tests, existing signature verification tests, and the new tests, etc.—go away, because we'd be adding tests only for the parsing. That is, this would allow us to “test along axes” instead of testing cross-products, and/or it would make doing the cross-product tests realistic performance-wise since no expensive crypto operations would be involved.

Further, imagine that we did change this PR to expose a `parse_spki()` instead of a new `verify()` function. Note that the `parse_spki()` function wouldn't be doing any cryptography; it's just doing parsing. So, does it really make sense for it to be in *ring*, if it isn't cryptography? Now I'm thinking that I should create a new crate that contains the contents of `ring::der` + the aforementioned `parse_spki()` function. Not only would this solve the issue, but it would also resolve the longstanding issue where webpki is using a private (`#[doc(hidden)]`) API: `ring::der`.

 WDYT?

marshallpierce commented:
That plan makes a lot of sense. Just looking at the functions in `ring::der`, they're all general purpose, and if webpki needs them too, that lends weight to pulling them out. A minor wrinkle is how to organize things like `parse_spki()` compared to the more general stuff in `der`, but no doubt a clear module structure can handle that. Going back to my original need, it would have solved my problem if I could easily massage an SPKI key into the format `signature::verify()` already knows about.

The one challenge with the plan is where to put the almost-constants like this that currently reside in `spki.rs`:

```
pub static RSA_PKCS1_2048_8192_SHA256: Algorithm = Algorithm {
    public_key_alg_id: RSA_ENCRYPTION,
    verification_alg: &signature::RSA_PKCS1_2048_8192_SHA256,
};
```

They don't fit into a `der` focused crate, but it's nice to do this for users to avoid having them use the wrong pub key algorithm ids, etc. This currently consists of the 3 der files in `src/signature/data`, a dependency on `ring` for the `VerificationAlgorithm`s, and the appropriate wiring of which key algorithm ids can be used with which verification algorithms. Maybe that should be its own ring spki crate that depends on the hypothetical der crate? It's essentially a bunch of handy constants.

briansmith commented:
OK, let's change this PR to expose `ring::signature::spki::parse_spki()` instead of `ring::signature::spki::verify()`, and then I'll do some experiments about factoring it out into its own crate. That way we don't delay getting something for you and others to use.

briansmith commented:
Also, I see this PR branch is based on a quite old branch.. Please *rebase* all the commits on top of master.

marshallpierce commented:
OK, I'll make those changes. Also, though the early commits have old timestamps, they are based on top of a891756d8df3a82d56c15f5752244df7a80fc2e8.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+#[derive(Debug)]
+pub struct SubjectPublicKeyInfo<'a> {
+    /// The algorithm id ASN.1.
+    pub algorithm_id_value: untrusted::Input<'a>,
```
I made this field public because it seemed like it might be generally useful.

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+/// Groups an ASN.1 AlgorithmIdentifier and a `ring` VerificationAlgorithm.
+pub struct Algorithm {
+    /// The `algorithm` member in SPKI from https://tools.ietf.org/html/rfc5280#section-4.1.
+    public_key_alg_id: AlgorithmIdentifier,
```
This one I didn't make public because I'm not sure why anyone outside this library would need it. But on the other hand, it feels odd to have one private field in a public struct...

marshallpierce commented on src/signature/spki.rs:
```diff
@@ -0,0 +1,202 @@
+    /// The `algorithm` member in SPKI from https://tools.ietf.org/html/rfc5280#section-4.1.
+    public_key_alg_id: AlgorithmIdentifier,
+    /// The verification algorithm corresponding to the algorithm id.
+    pub verification_alg: &'static signature::VerificationAlgorithm,
```
This one I made public because now people need to access it to use with the result of `parse_spki`.

marshallpierce commented on src/der.rs:
```diff
@@ -277,4 +317,11 @@ mod tests {
+    test_parse_bad_spki_der!(test_rsa_pkcs1_sha1_bad_key_der_length,
+                               "rsa-pkcs1-sha1-bad-key-der-length.pem");
+    test_parse_bad_spki_der!(test_rsa_pkcs1_sha1_bad_key_der_null,
+                               "rsa-pkcs1-sha1-bad-key-der-null.pem");
```
These are failing on android because the files are not present. I'll see if I can track down why.

marshallpierce replied:
Looks like `travis.sh` needs to copy over the chromium `*.pem` files like it copies over the `txt` test files. However, the current setup also isn't running integration tests (like the new spki ones) since it only copies over `target/(release|debug)/ring-*` and the test file for spki is `target/(release|debug)/signature_spki_tests-*`.

marshallpierce replied:
Android tests are now running all test targets.

briansmith commented:
Hi, I'm getting caught up on my PRs now and Id like to review and merge this but I'm not certain what the state of this is. I think we had some agreement to replace `ring::signature::spki::verify()` with a `parse_spki()` function, but the current PR doesn't seem to do that.

Please rebase the changes on top of the current master (I just tested it myself locally and there was only one conflict). Also, please clean up the commits by folding fixes into the commits they fix. For example, fold the "Fix a comment typo;" commit into the commit that introduced the typo. Similarly, instead of having a commit that introduces the new `verify()` function and then a commit that changes it to `parse_spki()`, let's just go straight to `parse_spki()`.

I know that rewriting the PRs is a PITA (I do it all the time), but it's going to be too hard and time consuming for me to review the PR in its current state, and the harder it is to review the longer it will take me to find time to merge it.

Thanks!

marshallpierce commented:
Perhaps your local clone is out of date, or I'm misunderstanding what you're after? `parse_spki` is there: https://github.com/briansmith/ring/pull/378/files#diff-b66c7d2261ee98e3b53d332a09c4e037R49

I've rebased and squashed so that the interstitial form of `verify` no longer exists.

marshallpierce commented:
Sorry I haven't gotten to this; I probably won't have time to go through your changes any time soon. I'm happy to let you do what you like with this PR's code as you see fit to make it usable for webpki, etc.

briansmith commented:
> Sorry I haven't gotten to this; I probably won't have time to go through your changes any time soon. I'm happy to let you do what you like with this PR's code as you see fit to make it usable for webpki, etc.

Thanks for your work on this. I'm going to close this issue and use what we learned from this to implement the feature later.
